File: src/TON_UPDATES/python_code/src/tests/checkrand.py
"""
checkrand.py - 1:1 Python port of checkrand.m

Stand-alone sanity check of the random policy with a synthetic penalty.
"""
import numpy as np

from _bootstrap import paths   # noqa: F401  (bootstraps sys.path)


def main():
    M = 10
    N = 10
    km = 2
    B = 20
    T = 10

    n = np.ones((M, km))
    c = np.ones(M)
    w = np.zeros((M, km))
    for m in range(M):
        for j in range(km):
            w[m, j] = (m + 1) / km

    K = T
    gamma = 0.9

    p = np.zeros((km, B))
    p[0, :B] = np.log(np.arange(1, B + 1))
    p[1, :B] = np.arange(1, B + 1)

    Delta = np.zeros((M, km), dtype=np.int64)
    pavg = np.zeros(K)
    presult = 0.0

    for t in range(K):
        g = np.zeros((M, km))
        Ccurr = np.zeros(M)
        Ncurr = 0.0
        Change = np.zeros((M, km), dtype=bool)

        source = np.random.permutation(M)[:min(N, M)]
        # MATLAB while-loop: fill until Ncurr >= N, indexing into `source`
        i = 0
        while Ncurr < N and i < len(source):
            row = int(source[i])
            # MATLAB randperm(km, min(c(row), km)) -> random subset of tasks
            n_pick = int(min(c[row], km))
            columnrand = np.random.permutation(km)[:n_pick]
            for column in columnrand:
                g[row, column] = 1
                Delta[row, column] = 0
                Ccurr[row] += 1
                Ncurr += n[row, column]
                Change[row, column] = True
            i += 1

        for m in range(M):
            for j in range(km):
                if not Change[m, j]:
                    if Delta[m, j] + 1 > B - 1:
                        Delta[m, j] = B - 1
                    else:
                        Delta[m, j] = Delta[m, j] + 1
                pavg[t] += (w[m, j] * p[j, Delta[m, j]]) / (km * M)

        if t + 1 < T:
            presult += (gamma ** t) * pavg[t]

    print(f"presult = {presult}")


if __name__ == "__main__":
    main()

File: src/TON_UPDATES/matlab_code/checkrand.m
close all;
clear all;
clc;

%% number of sources, channels, bound, and time
M=10;
N=10;
km=2;
B=20;
T=10;
%% channel needed
for m=1:M
for j=1:km
    if j>km/2
       n(m,j)=1;
    else
        n(m, j)=1;
    end
end
end
%% computation resources
for m=1:M
    if m>M/2
        c(m)=1;
    else
        c(m)=1;
    end
end
%% different weight
w=zeros(1,M);
for m=1:M
    for j=1:km
      w(m, j)=m/km;
    end
end
K=T;
gamma=0.9;
%% penalty function
delta=1:B;
% loss=load('loss.mat', 'p1', 'p2');
% p=zeros(km,B);
% p1=loss.p1;
% p2=loss.p2;
% i=1;
% for j=1:5:100
%     p(1,i)=p1(j);
%     p(2,i)=0.01*p2(j);
%     i=i+1;
% end
p(1,1:B)=log(1:B);
p(2,1:B)=1:B;

%% initialization
Delta=ones(M,km); % State initialization
pavg=zeros(1,K);
presult=0;
for t=1:K
    %% Initialization
    g=zeros(M,km);
    Ccurr=zeros(1,M);
    Ncurr=0;
    Change=zeros(M,km);
    source=randperm(M,min(N,M));
    i=1;
    while Ncurr<N
        row=source(i);
        columnrand=randperm(km,min(c(row),km));
        for j=1:length(columnrand)
            column=columnrand(j);
              g(row, column)=1;
              Delta(row, column)=1;
              Ccurr(row)=Ccurr(row)+1;
              Ncurr=Ncurr+n(row, column);
              Change(row, column)=1;
        end
        i=i+1;
    end

    %% update AoI
    for m=1:M
        for j=1:km
            if Change(m,j)==0
                if Delta(m, j)+1>B
                    Delta(m, j)=B;
                else
                    Delta(m, j)=Delta(m, j)+1;
                end
            end
           pavg(t)=pavg(t)+(w(m, j)*p(j,Delta(m,j)))/(km*M);
        end
    end
   
    if t<T
    presult=presult+gamma^(t-1)*pavg(t);
    end
end
presult